pkg/experiment/symbolizer/symbolizer.go

```diff
@@ -0,0 +1,448 @@
+	if mapping.Filename < 0 || int(mapping.Filename) >= len(profile.StringTable) {
+		return "", fmt.Errorf("invalid mapping: filename index %d out of range (string table length: %d)",
+			mapping.Filename, len(profile.StringTable))
+	}
+
+	fullPath := profile.StringTable[mapping.Filename]
+	return filepath.Base(fullPath), nil
+}
+
+// extractBuildID extracts and sanitizes the build ID from the mapping
+func (s *Symbolizer) extractBuildID(profile *googlev1.Profile, mapping *googlev1.Mapping) (string, error) {
+	buildID := profile.StringTable[mapping.BuildId]
+	sanitizedBuildID, err := sanitizeBuildID(buildID)
+	if err != nil {
+		level.Error(s.logger).Log("msg", "Invalid buildID", "buildID", buildID)
+		return "", err
+	}
+
+	return sanitizedBuildID, nil
+}
+
+// createSymbolizationRequest creates a symbolization request for a mapping group
+func (s *Symbolizer) createSymbolizationRequest(binaryName, buildID string, mapping *googlev1.Mapping, locs []locToSymbolize) request {
+	req := request{
+		buildID:    buildID,
+		binaryName: binaryName,
+		locations:  make([]*location, len(locs)),
+	}
+
+	for i, loc := range locs {
+		req.locations[i] = &location{
+			address: loc.loc.Address,
+			mapping: &pprof.Mapping{
+				Start:   mapping.MemoryStart,
+				Limit:   mapping.MemoryLimit,
+				Offset:  mapping.FileOffset,
+				BuildID: buildID,
+			},
+		}
+	}
+
+	return req
+}
+
+// updateProfileWithSymbols updates the profile with symbolization results
+func (s *Symbolizer) updateProfileWithSymbols(profile *googlev1.Profile, mapping *googlev1.Mapping, locs []locToSymbolize, symLocs []*location) {
+	stringMap := make(map[string]int64, len(profile.StringTable))
+	for i, str := range profile.StringTable {
+		stringMap[str] = int64(i)
+	}
+
+	type funcKey struct {
+		nameIdx, filenameIdx int64
+	}
+	funcMap := make(map[funcKey]uint64)
+	maxFuncID := uint64(0)
+
+	for _, fn := range profile.Function {
+		if fn.Id > maxFuncID {
+			maxFuncID = fn.Id
+		}
+		funcMap[funcKey{fn.Name, fn.Filename}] = fn.Id
+	}
+
+	for i, symLoc := range symLocs {
+		locIdx := locs[i].idx
+		profile.Location[locIdx].Line = make([]*googlev1.Line, len(symLoc.lines))
+
+		for j, line := range symLoc.lines {
+			nameIdx, ok := stringMap[line.FunctionName]
+			if !ok {
+				nameIdx = int64(len(profile.StringTable))
+				profile.StringTable = append(profile.StringTable, line.FunctionName)
+				stringMap[line.FunctionName] = nameIdx
+			}
+
+			filenameIdx, ok := stringMap[line.FilePath]
+			if !ok {
+				filenameIdx = int64(len(profile.StringTable))
+				profile.StringTable = append(profile.StringTable, line.FilePath)
+				stringMap[line.FilePath] = filenameIdx
+			}
+
+			key := funcKey{nameIdx, filenameIdx}
+			funcID, ok := funcMap[key]
+			if !ok {
+				maxFuncID++
+				funcID = maxFuncID
+				profile.Function = append(profile.Function, &googlev1.Function{
+					Id:        funcID,
+					Name:      nameIdx,
+					Filename:  filenameIdx,
+					StartLine: int64(line.LineNumber),
```

Comment:
lets remove this line. On the left we have the start of the function and on the right we have the middle of the function. Also the line number is not included in the key. Also we don't really have line numbers at the moment.

Reply:
we also don' really have filenames yet. I suggest we remove filename here as well.